### ISSUE TITLE
Linear RAM memory increase with Dataset and Estimator with epoch loops

### ISSUE THREAD

**Issue by Mageswaran1989** [tensorflow/tensorflow#32052]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.2 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below):  tensorflow-gpu==2.0.0-rc0
- Python version: 3.7.1
- Bazel version (if compiling from source): NA
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version: Cuda 10.1 / cuDNN 7.6
- GPU model and memory: GeForce GTX 1060 Mobile 6GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When using Dataset with Estimator, the memory foot print of RAM keeps raising when estimator's train and evaluate APIs are called in loop.

**Describe the expected behavior**
RAM usage should not increase with epochs.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
Please find the source code @ https://gist.github.com/Mageswaran1989/facc3fc2a003807d029a914c721629db

[Update] My latest test case @ https://github.com/dhiraa/tf_issue_32052

StackOverflow Rereference :  https://stackoverflow.com/questions/55211315/memory-leak-with-tf-data

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

[Updated the graph]
![tf_memory_test](https://user-images.githubusercontent.com/3304549/63917881-30bb6b80-ca59-11e9-8598-ebdca798dff5.png)

**Comment by Mageswaran1989** [comment]
@goldiegadde Thanks for the quick response.

The [Gist script](https://gist.github.com/Mageswaran1989/facc3fc2a003807d029a914c721629db) I posted here replicates my setup with dummy data and dummy model.

When actual image data and heat map used with Resnet50 model architecture, the memory increase slowly for each epoch and eats up the whole RAM.
You can find the graph from my setup, for 10 epochs. If I train with 100 epochs it gets killed by OS in the middle.

![east_memory](https://user-images.githubusercontent.com/3304549/63904650-64819b80-ca2f-11e9-85aa-dc9ff7217d7c.png)

I even tried calling explicit `gc.collect()`, but no use.

If time permits I will put up a small repo with actual data and model.

**Comment by Mageswaran1989** [comment]
similar issue @ https://github.com/tensorflow/tensorflow/issues/24047

**Comment by jnd77** [comment]
Not sure if it helps, but I had a similar issue. I solved it with installing and running tcmalloc allocator:

https://github.com/tensorflow/tensorflow/issues/2942

**Comment by Mageswaran1989** [comment]
@goldiegadde  Please find the attached zip file, which has the README to reproduce the issue.

[tf_dataset_memory_test.zip](https://github.com/tensorflow/tensorflow/files/3555207/tf_dataset_memory_test.zip)

With simple model , as you pointed out there is not much significant difference with each epoch.

However with convolutional networks, complex loss calculation and optimization routines, the memory footprint is very large.

Note : The EAST model in the zip file is still under porting and testing phase from TF 1.x version. It may also have something to do with this memory increase.

**Comment by Mageswaran1989** [comment]
Some findings with [mem_profiler](https://pypi.org/project/memory-profiler/). After each dataset function call the memory increases.

```python
# epoch 1

Line #    Mem usage    Increment   Line Contents
================================================
   398    302.4 MiB    302.4 MiB   @profile
   399                             def train(estimator, max_steps=None):
   400    302.5 MiB      0.1 MiB       train_spec = _get_train_spec(max_steps=max_steps)
   401    302.5 MiB      0.0 MiB       estimator.train(
   402    302.5 MiB      0.0 MiB           input_fn=train_spec.input_fn,
   403    302.5 MiB      0.0 MiB           hooks=train_spec.hooks,
   404   5257.2 MiB   4954.7 MiB           max_steps=train_spec.max_steps)


Line #    Mem usage    Increment   Line Contents
================================================
   407   5257.0 MiB   5257.0 MiB   @profile
   408                             def evaluate(estimator, steps=None, checkpoint_path=None):
   409   5257.0 MiB      0.0 MiB       eval_spec = _get_eval_spec(steps=steps)
   410   5257.0 MiB      0.0 MiB       estimator.evaluate(
   411   5257.0 MiB      0.0 MiB           input_fn=eval_spec.input_fn,
   412   5257.0 MiB      0.0 MiB           steps=eval_spec.steps,
   413   5257.0 MiB      0.0 MiB           hooks=eval_spec.hooks,
   414   7506.0 MiB   2249.0 MiB           checkpoint_path=checkpoint_path)

#epoch 2

Line #    Mem usage    Increment   Line Contents
================================================
   398   7506.2 MiB   7506.2 MiB   @profile
   399                             def train(estimator, max_steps=None):
   400   7506.2 MiB      0.0 MiB       train_spec = _get_train_spec(max_steps=max_steps)
   401   7506.2 MiB      0.0 MiB       estimator.train(
   402   7506.2 MiB      0.0 MiB           input_fn=train_spec.input_fn,
   403   7506.2 MiB      0.0 MiB           hooks=train_spec.hooks,
   404   9878.6 MiB   2372.4 MiB           max_steps=train_spec.max_steps)

Line #    Mem usage    Increment   Line Contents
================================================
   407   9878.6 MiB   9878.6 MiB   @profile
   408                             def evaluate(estimator, steps=None, checkpoint_path=None):
   409   9878.6 MiB      0.0 MiB       eval_spec = _get_eval_spec(steps=steps)
   410   9878.6 MiB      0.0 MiB       estimator.evaluate(
   411   9878.6 MiB      0.0 MiB           input_fn=eval_spec.input_fn,
   412   9878.6 MiB      0.0 MiB           steps=eval_spec.steps,
   413   9878.6 MiB      0.0 MiB           hooks=eval_spec.hooks,
   414  12068.0 MiB   2189.4 MiB           checkpoint_path=checkpoint_path)

```

**Comment by burness** [comment]
@Mageswaran1989  I met the same issue, have you solve it  ?

**Comment by Mageswaran1989** [comment]
@burness nope. We faced similar issue while using generator with TF dataset, we thought we were doing something wrong in our generator code. So we adopted different strategy. 

Can you share your setup details ? which may give us more clues

**Comment by Mageswaran1989** [comment]
@jsimsa Please find the current version of the code I use to replicate this memory issue @ https://github.com/dhiraa/tf_issue_32052

**Comment by jsimsa** [comment]
ymodak@ the memory increase happens inside of `estimator.train` and `estimator.evaluate` and is not necessarily coming from tf.data. As per user's previous instructions, using a similar input pipeline with a trivial model does not reproduce the memory increase. So someone on the high-level APIs / Estimator team should pinpoint where is the memory increase coming from before and if it is indeed coming from tf.data, I would be happy to take a further look.

**Comment by burness** [comment]
@Mageswaran1989 Thanks for your reply,   I found that if I use "prefetch" after the parallel_interleave，the code which use estimator to train model will increase RAM memory, but If I just read dataset from hdfs, it's ok.

**Comment by Mageswaran1989** [comment]
No luck when I disabled `prefetch` on my code

**Comment by Mageswaran1989** [comment]
I get this warning 

```python
 Entity <bound method CodeMap.items of {<code object east_features_decode at 0x7f0b09e84e40, file "/opt/tf_issue_32052/dummy_datasets.py", line 139>: {}}> appears to be a generator function. It will not be converted by AutoGraph.

```

**Comment by Mageswaran1989** [comment]
When I simply iterate through the dataset I could see the memory increase.

[EAST Dataset : east_itr_log.txt](https://github.com/dhiraa/tf_issue_32052/blob/master/east_itr_log.txt) 
[Numpy Dataset: numpy_itr_log.txt](https://github.com/dhiraa/tf_issue_32052/blob/master/numpy_itr_log.txt)
[Code](https://github.com/dhiraa/tf_issue_32052/blob/master/dummy_datasets.py#L208)

Please check my [repo](https://github.com/dhiraa/tf_issue_32052) for test command

**Comment by Mageswaran1989** [comment]
I dont know whether this is a valid test case, however when I create a [3-D random array](https://github.com/dhiraa/tf_issue_32052/blob/master/dummy_datasets.py#L144) as part of `dataset map` operation, the memory increases.

Also this memory consumption is only once, even if I call iterator twice the memory remains same.

```python
def numpy_array_decode(serialized_example,
                       NUM_FEATURES=250): #TODO make it as arg
    # define a parser
    features = tf.io.parse_single_example(
        serialized_example,
        # Defaults are not specified since both keys are required.
        features={
            'data': tf.io.FixedLenFeature([1 * NUM_FEATURES], tf.float32),
            'label': tf.io.FixedLenFeature([1], tf.float32),
        })

    data = tf.reshape(
        tf.cast(features['data'], tf.float32), shape=[1, NUM_FEATURES])

    label = tf.reshape(
        tf.cast(features['label'], tf.float32), shape=[1])

     return {"data": data, "dummy": np.random.rand(512, 512, 5)}, label
    #return {"data": data}, label

```

**Comment by caissalover** [comment]
I'm too facing a similar issue

My code is something like
        `test = [None]*50`
        `for i in range(50):`
        `    test[i] = fooWorker( numpyData, constantTensors, i)`

where fooWorker is my customOp decorated with @tf.function. It returns some small lists and I just don't get why memory grows and is not collected in garbage. numpyData is substantial in size, all the other sizes are minimal.

It seems similar to case in tf1 where we kept creating graphs instead of simulating in sess.run

**Comment by aaudiber** [comment]
@Mageswaran1989 If you define the dataset just once, then re-iterate over it during each epoch, it will reduce the memory consumption. Otherwise each time you create a new dataset, the dataset will be added to the global default graph, so it won't be cleaned up when it leaves scope.

```python
dataset = files.interleave(...)
for i in range(num_epochs):
  for elem in dataset:
    # process elem
```

Creating the dataset once and then iterating over it once per epoch is the idiomatic way of doing things in TF2.

**Comment by Mageswaran1989** [comment]
@aaudiber Thanks for clearing my assumption. As I thought the data is going to the global graph.  

I tried your suggestion before raising this issue, when I came across this [StackOverflow question](https://github.com/tensorflow/tensorflow/issues/32052) . The problem arises when I use dataset with estimator.

I get following error, when I try to move the dataset creation outside of `input_fn` definition.

```python
The graph (<tensorflow.python.framework.ops.Graph object at 0x7f2837489898>) of the iterator is different from the graph (<tensorflow.python.framework.ops.Graph object at 0x7f283748b278>) the dataset: Tensor("PrefetchDataset:0", shape=(), dtype=variant) was created in. If you are using the Estimator API, make sure that no part of the dataset returned by the `input_fn` function is defined outside the `input_fn` function. Please ensure that all datasets in the pipeline are created in the same graph as the iterator. NOTE: This warning will become an error in future versions of TensorFlow.
```

 If I keep it part of the `input_func` and whenever [`tf.estimator.TrainSpec`](https://www.tensorflow.org/api_docs/python/tf/estimator/TrainSpec) is getting called by the Estimator there is a memory increase.

**Comment by jsimsa** [comment]
The problem is that you are calling estimator `train` method multiple times (once for each epoch). This will create multiple copies of the input pipeline graph (and since the input pipeline graph inlines any constants such as input to `from_tensors` and `from_tensor_slices`, you will see a memory increase.

The recommended way for iterating through multiple epochs of your data using the Estimator API is to use the `tf.data.Dataset.repeat` transformation in your input pipeline.

**Comment by Mageswaran1989** [comment]
@jsimsa I just double confirmed in my setup, even if I use `repeat` along with step calculation, the RAM memory that gets allocated for my data size is out of proportion.

 **Numpy Datasize : ~ 52MB + ~21MB**

```python
Line #    Mem usage    Increment   Line Contents
================================================
   149    261.8 MiB    261.8 MiB   @profile
   150                             def train_n_evaluate(estimator, TRAIN_DATA, VAL_DATA, BATCH_SIZE, IS_EAST_IMAGE_TEST, max_steps=None, NUM_EPOCHS=None):
   151    261.8 MiB      0.0 MiB       train_spec = _get_train_spec(TRAIN_DATA=TRAIN_DATA,
   152    261.8 MiB      0.0 MiB                                    max_steps=max_steps,
   153    261.8 MiB      0.0 MiB                                    BATCH_SIZE=BATCH_SIZE,
   154    261.8 MiB      0.0 MiB                                    IS_EAST_IMAGE_TEST=IS_EAST_IMAGE_TEST,
   155    439.8 MiB    178.0 MiB                                    NUM_EPOCHS=NUM_EPOCHS)
   156    439.8 MiB      0.0 MiB       eval_spec = _get_eval_spec(VAL_DATA=VAL_DATA,
   157    439.8 MiB      0.0 MiB                                  steps=200,
   158    439.8 MiB      0.0 MiB                                  BATCH_SIZE=BATCH_SIZE,
   159    439.8 MiB      0.0 MiB                                  IS_EAST_IMAGE_TEST=IS_EAST_IMAGE_TEST)
   160                             
   161  11630.0 MiB  11190.2 MiB       tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
```

**Image Datasize : ~140 MB + ~28MB**

```python

Line #    Mem usage    Increment   Line Contents
================================================
   149    261.8 MiB    261.8 MiB   @profile
   150                             def train_n_evaluate(estimator, TRAIN_DATA, VAL_DATA, BATCH_SIZE, IS_EAST_IMAGE_TEST, max_steps=None, NUM_EPOCHS=None):
   151    261.8 MiB      0.0 MiB       train_spec = _get_train_spec(TRAIN_DATA=TRAIN_DATA,
   152    261.8 MiB      0.0 MiB                                    max_steps=max_steps,
   153    261.8 MiB      0.0 MiB                                    BATCH_SIZE=BATCH_SIZE,
   154    261.8 MiB      0.0 MiB                                    IS_EAST_IMAGE_TEST=IS_EAST_IMAGE_TEST,
   155    439.8 MiB    178.0 MiB                                    NUM_EPOCHS=NUM_EPOCHS)
   156    439.8 MiB      0.0 MiB       eval_spec = _get_eval_spec(VAL_DATA=VAL_DATA,
   157    439.8 MiB      0.0 MiB                                  steps=200,
   158    439.8 MiB      0.0 MiB                                  BATCH_SIZE=BATCH_SIZE,
   159    439.8 MiB      0.0 MiB                                  IS_EAST_IMAGE_TEST=IS_EAST_IMAGE_TEST)
   160                             
   161  11630.0 MiB  11190.2 MiB       tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)

```

Reference Log: 
[5epochs_with_repeat.log](https://github.com/tensorflow/tensorflow/files/3585336/5epochs_with_repeat.log)

**Comment by aaudiber** [comment]
@Mageswaran1989 can you simplify your repro example as much as possible, ideally in a single file that demonstrates the issue? That would be super helpful for finding the root cause.

**Comment by Mageswaran1989** [comment]
@aaudiber Please find the updated file @ https://github.com/dhiraa/tf_issue_32052. Hope it helps!

Observations 
- Memory remains constant when loop is not used, however the memory it takes is disproportional 
- Previously when I used loop, I used to run evaluation after each epoch and store the evaluation logs, in current `train_and_evaluate`  setup I had to run the entire `num_epochs` before I can `evaluate` my model. I may be missing something here or may be some API can help me out with this.


**Logs**
```python


Filename: test_memory_leak.py

Line #    Mem usage    Increment   Line Contents
================================================
   367    300.7 MiB    300.7 MiB   @profile
   368                             def train_n_evaluate(estimator,
   369                                                  train_data_path,
   370                                                  val_data_path,
   371                                                  batch_size,
   372                                                  num_features,
   373                                                  num_epochs=None,
   374                                                  max_train_steps=None,
   375                                                  max_val_steps=None):
   376    300.7 MiB      0.0 MiB       train_spec = _get_train_spec(train_data_path=train_data_path,
   377    300.7 MiB      0.0 MiB                                    batch_size=batch_size,
   378    300.7 MiB      0.0 MiB                                    num_features=num_features,
   379    300.7 MiB      0.0 MiB                                    num_epochs=num_epochs,
   380    420.7 MiB    119.9 MiB                                    max_steps=max_train_steps)
   381    420.7 MiB      0.0 MiB       eval_spec = _get_eval_spec(val_data_path=val_data_path,
   382    420.7 MiB      0.0 MiB                                  batch_size=batch_size,
   383    420.7 MiB      0.0 MiB                                  num_features=num_features,
   384    420.7 MiB      0.0 MiB                                  num_epochs=num_epochs,
   385    420.7 MiB      0.0 MiB                                  max_steps=max_val_steps)
   386                             
   387  10256.1 MiB   9835.4 MiB       tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)


Filename: test_memory_leak.py

Line #    Mem usage    Increment   Line Contents
================================================
   407    242.0 MiB    242.0 MiB   @profile
   408                             def main(args):
   409    242.0 MiB      0.1 MiB       memory_usage_psutil("1. Before generating data")
   410                             
   411                                 #  1. Generate regression data
   412    242.0 MiB      0.0 MiB       generate_numpy_tf_records(out_dir=args["train_path"],
   413    242.0 MiB      0.0 MiB                                 num_tfrecord_files=args["num_tfrecord_train_files"],
   414    242.0 MiB      0.0 MiB                                 num_samples_per_file=args["num_samples_per_file"],
   415    262.2 MiB     20.2 MiB                                 num_features=args["num_features"])
   416    262.2 MiB      0.0 MiB       generate_numpy_tf_records(out_dir=args["val_path"],
   417    262.2 MiB      0.0 MiB                                 num_tfrecord_files=2,
   418    262.2 MiB      0.0 MiB                                 num_samples_per_file=args["num_samples_per_file"],
   419    300.5 MiB     38.3 MiB                                 num_features=args["num_features"])
   420                             
   421    300.5 MiB      0.0 MiB       total_steps_per_file = args["num_samples_per_file"] / args["batch_size"]
   422                             
   423    300.5 MiB      0.0 MiB       memory_usage_psutil("2. Before defining model")
   424                             
   425                                 # 2. Define the model
   426    300.5 MiB      0.0 MiB       model = NNet()
   427                             
   428    300.5 MiB      0.0 MiB       memory_usage_psutil("3. Before defining estimator")
   429                             
   430                                 # 3. Define engine to train i.e Estimator
   431    300.5 MiB      0.0 MiB       estimator = tf.estimator.Estimator(model_fn=model,
   432    300.5 MiB      0.0 MiB                                          config=_init_tf_config(total_steps_per_file=total_steps_per_file,
   433    300.7 MiB      0.2 MiB                                                                 model_dir=args["model_dir"]),
   434    300.7 MiB      0.0 MiB                                          params=None)
   435                             
   436    300.7 MiB      0.0 MiB       memory_usage_psutil("4. Before training")
   437                             
   438                                 # 4. Train and evaluate the model with generated regression data
   439    300.7 MiB      0.0 MiB       train_n_evaluate(estimator=estimator,
   440    300.7 MiB      0.0 MiB                        train_data_path=args["train_path"],
   441    300.7 MiB      0.0 MiB                        val_data_path=args["val_path"],
   442    300.7 MiB      0.0 MiB                        batch_size=args["batch_size"],
   443    300.7 MiB      0.0 MiB                        num_features=args["num_features"],
   444    300.7 MiB      0.0 MiB                        num_epochs=args["num_epochs"],
   445    300.7 MiB      0.0 MiB                        max_train_steps=None,
   446  10256.1 MiB   9955.3 MiB                        max_val_steps=None)
   447                             
   448  10256.1 MiB      0.0 MiB       memory_usage_psutil("5. Before exporitng the model")
   449                             
   450  10266.6 MiB     10.5 MiB       export_model(estimator=estimator, model_export_path=args["model_export_path"], num_features=args["num_features"])

INFO:absl:>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
WARNING:absl:Final memory usage:  : Memory used is 10266.5546875
INFO:absl:<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
```

**Comment by aaudiber** [comment]
Thanks @Mageswaran1989. It looks like this follows the expected behavior that memory increases linearly only when the loop is outside of Dataset/Estimator. You should be able to achieve everything you need without having an external loop. 

To perform evaluation after every epoch, remove the `.repeat()` from your dataset input function. See [this](https://stackoverflow.com/a/49655277/401884) stackoverflow answer.

**Comment by Mageswaran1989** [comment]
As per the stackoverflow answer, I tried the dataset to repeat once (as well as removing dataset repeat) and used calculated `max_steps`  in the train spec.

Expectations/My assumption:
- When train dataset reaches its end, the training paused and evaluation kicks in.
- After evaluation is done, the training starts again 

Observed:
- Only once the evaluation is done irrespective of dataset settings. Estimator evaluation starts only after the train `max_steps` are reached

**Comment by arthurflor23** [comment]
I had this problem just using ``model.fit_generator()`` on 2.0.0-rc1, so I back to the 2.0.0-beta1

**Comment by Mageswaran1989** [comment]
Can anyone help me with the right way of setting up the TFEstimator and Dataset APIs in Tensorflow 2.0 so that I can evaluate after each epoch, like we did in Tensorflow 1.1x ?

**Comment by jsimsa** [comment]
@karmel could you please comment on the recommended use of Estimator APIs in TF 2.0?

**Comment by dlutkaka** [comment]
same issue,
same code will cause gpu memory leak on tf-gpu 2.0.0-rc-0 and 2.0.0-rc-1, but work fine on 2.0.0beta1

**Comment by ialdencoots** [comment]
I'm not sure if it's related to this same issue, but just iterating through a regular `TFRecordDataset` and using `model.predict_on_batch` results in a gpu memory leak that eventually crashes with an OOM exception for me.

**Comment by Callahman** [comment]
Not sure if this is the same issue, but I'm experiencing linear memory leaks on loops as well [example below]. I tried including a recommended fix from other threads (k.clear_session()), but that only slowed the memory leak.

Let me know if I should open a separate ticket. Hope this helps/a fix can be found.

OS: Ubuntu 18.04.3
GPU: GeForce GTX 1080/PCIe/SSE2

`import tensorflow as tf
from tensorflow.keras import layers
import numpy as np
import psutil
import matplotlib.pyplot as plt

from tensorflow.keras import backend as k



model = tf.keras.Sequential()
model.add(layers.Dense(4, input_shape = (50,)))
model.compile(loss='mse')


x = []
y = []
for xs in range(10000):
    
    values = np.random.uniform(size = (1,50))
    prediction = model.predict(values)
    
    k.clear_session()
    
    mem = psutil.virtual_memory()
    mem = mem[2]
    
    x.append(xs)
    y.append(mem)
    
    
plt.plot(x, y)`

**Comment by Mageswaran1989** [comment]
any updates?

**Comment by FelixHo** [comment]
same issue, TF 2.0
I run Estimator with distributed mode on 6 machines ( PS x 1 / evaluator x 1 / worker x 4) ,
my code is super simple：
```
model = tf.estimator.DNNLinearCombinedClassifier()
train_spec = tf.estimator.TrainSpec(...) # input_fn: epoches = infinite
eval_spec  = tf.estimator.EvalSpec(...) # input_fn: epoches = 1 
tf.estimator.train_and_evaluate(model, train_spec, eval_spec)
```
As `save_checkpoints_steps = 500` ,  so it will generate a checkpoint every 500 steps, and the evaluator starts to evaluate. I found that each evaluation will result in an increase of about 500MB of memory used on the evaluator node and other nodes (PS & Worker) did not change significantly. 

PS: **Even if I use very small datasets when evaluating, memory leaks still exist. (~500MB↑/run)**

**Comment by dlutkaka** [comment]
using keras api
calling model.evaluate() in "on_batch_end()" callback will lead memory leaks

**Comment by karmel** [comment]
Sorry for the delay. There is a lot going on in this thread, and many of these issues are unrelated.

@Mageswaran1989 -- Estimator is not designed to interleave calls to train and evaluate (or predict), and we would recommend the train_and_evaluate API. If you want to be able to interleave and inspect between epochs, I recommend looking at the [Keras](https://www.tensorflow.org/guide/keras) API instead. This will be more 2.0 native, and more conducive to fine-grained control of the training loop.

@ialdencoots , @Callahman , @dlutkaka -- these are unrelated problems. If you are still having issues, please open new Issues with code samples and the information requested in the bug template.

@FelixHo -- this may or may not be a related issue, but it's hard to tell without more information. Please file a new issue with a minimal code sample + the information requested in the bug template.

Thanks, all.

**Comment by Danmou** [comment]
I was having a similar issue with both TF 2.0 and nightly using the Keras API to fit the model while doing some predictions between epochs with a callback. TF was leaking memory quite rapidly (a rate of around 8GB/hr), making long training runs impossible. I tried using [tcmalloc](http://goog-perftools.sourceforge.net/doc/tcmalloc.html) as suggested in [this guide](https://dantkz.github.io/How-To-Debug-A-Memory-Leak-In-TensorFlow/) and it solved the problem completely. If you're using conda, the fix is as simple as installing the `gperftools` package from `conda-forge` and setting the `LD_PRELOAD` flag like so:
```
LD_PRELOAD=$CONDA_PREFIX/lib/libtcmalloc.so python your_training_script.py
```

**Comment by Corea** [comment]
Any more updates? I had similar issue on [Google Cloud AI Platform](https://cloud.google.com/ai-platform), and I cannot use different `tcmalloc` on the cloud. Memory leakage got worse especially when I concatenated multiple dataset.

**Comment by DanielFirmbach** [comment]
I too had a memory leak, where my RAM usage increased linearly. I could resolve it by removing the num_parallel_calls parameter of the map function when using it on my dataset.

**Comment by Corea** [comment]
Like @DanielFirmbach said, I also did remove `num_parallel_cells` of map function too. A memory leak was alleviated pretty much, but the problem still exists somewhere. Of course, training speed got worse than before about 20%. Here is graph of worker memories.


- When I use `num_parallel_cells` of map function on dataset.
![image](https://user-images.githubusercontent.com/1433166/74820096-c1a63c00-5344-11ea-8762-0f4ba4971188.png)


- After removing parameter.
![image](https://user-images.githubusercontent.com/1433166/74820103-c4a12c80-5344-11ea-9ec9-89420b26dc91.png)

**Comment by brucechou1983** [comment]
Same here in TF 2.1.0. Any updates?

**Comment by mrashidmaqbool** [comment]
I am facing an issue probably due to memory leak. 
I am using a Dataset generator which upload images in batches. If I only train images without validation, everything works fine. But as soon as I introduce validation dataset generator, i start facing an issue of memory. 

1. Only training data: no memory issue
2. training data and validation data: memory issue. 

I have one dataset generator which is called twice (one by training dataset and one by validation)

any solution to resolve this issue.

**Comment by yasinalm** [comment]
In my case, `input_signature` argument in `tf.function` decorated functions dramatically reduced memory usage.

For example, if you have a function like this:
```
def train_step(self, image_stack):
    # Forward pass with gradient tape and loss calc
    with tf.GradientTape() as tape:
      image_reconstructed = self.model(image_stack)
      loss = self.calc_loss(image_stack, image_reconstructed )

    # Obtain gradients of trainable vars w.r.t. loss and perform update
    gradients = tape.gradient(loss, self.model.trainable_weights)
    self.optimizer.apply_gradients(zip(gradients, self.model.trainable_weights))

    return (image_reconstructed, loss)

```

You can specify the `input_signature` like this:
```
self.train_step = tf.function(self.train_step,
    input_signature=[
        tf.TensorSpec(shape=[None, None, None, 3], dtype=tf.float32),
      ]
    )
```
where the input image is of the shape `[batch, height, width, channels]`, and of the type `tf.float32`.

Although this avoids expensive retracing and significantly reduces memory usage, I still have an increasing memory usage.

**Comment by Mageswaran1989** [comment]
This is observable when we try to design our code to train the model based on epochs fashion in a user loop.

**Comment by yasinalm** [comment]
I figured it out. I have no RAM OOM now. In my case, the problem was `Dataset.cache()`. If there are random transformations (e.g. images augmentations) in the dataset pipeline, they should be after `Dataset.cache()`. Otherwise, it looks like `Dataset.cache()` caches all possible random transformations.

**Comment by swghosh** [comment]
> Like @DanielFirmbach said, I also did remove `num_parallel_cells` of map function too. A memory leak was alleviated pretty much, but the problem still exists somewhere. Of course, training speed got worse than before about 20%. Here is graph of worker memories.
> 
> * When I use `num_parallel_cells` of map function on dataset.
>   ![image](https://user-images.githubusercontent.com/1433166/74820096-c1a63c00-5344-11ea-8762-0f4ba4971188.png)
> * After removing parameter.
>   ![image](https://user-images.githubusercontent.com/1433166/74820103-c4a12c80-5344-11ea-9ec9-89420b26dc91.png)

I'm facing a similar issue with TFv2.3 as well although, I'm using `keras.Model.fit` for training on a TPU. I'd assume that the same is happening due to use of `ds.map(map_fn, AUTOTUNE)` with AUTOTUNE?

PS: my map_fn makes use of a lot of Python dicts and lists which are serialized together with the help of two for loops executed with `tf.autograph`.

https://github.com/tanzhenyu/image_augmentation/blob/master/image_augmentation/image/augmentation.py#L402-L432
https://github.com/tanzhenyu/image_augmentation/blob/master/image_augmentation/ai_platform/train_efficientnet.py#L251-L252

Typically, I'd see that there is a linear increase in the memory consumption of the VM (**the VM which helps run model train** on TPU, not the CPU side of the TPU host) per 5-10 epochs. I do hope that the same is due to some form of an increase in the graph size of the dataset pipeline due to the clogging of Python objects/functions repeatedly. Training would bottleneck out after like 100 epochs.

/cc: @tanzhenyu

**Comment by NikZak** [comment]
Maybe similar workaround as in [here](https://github.com/google/automl/issues/688) would generally be useful. 

The idea is to run the content of the epoch loop in a child process and thus memory is being released after every epoch iteration.

**Comment by swghosh** [comment]
@NikZak I am not sure if we could use the same TPU device across different processes.

**Comment by NikZak** [comment]
@swghosh Valid concern. However, I am not sure whether it will be an issue. There is no parallelization here. One process is made per one epoch and epochs are run sequentially. Every time an epoch ends the process also ends. There are always only two simultaneously running processes - the parent process and the child process created inside the epoch loop. Does this help to address your concern?

**Comment by swghosh** [comment]
@NikZak I guess it should work with forking a child process every time for the evals. Just that it'd cost us a bit of overhead in connecting to the TPU device (in the child process) each time we want to perform validation in between training.

**Comment by hoodootechnology** [comment]
@Mageswaran1989 did you find a solution to this?

**Comment by Mageswaran1989** [comment]
Nope. I haven't returned back to TF after this memory issue. I was TF using from version 1.0 and the TF 2.x Keras changes made me to loose the track of things sadly.

**Comment by vvsotnikov** [comment]
I experience the same issue on 2.4.0.

**Comment by necronomican** [comment]
> I experience the same issue on 2.4.0.

Same

**Comment by AlbertoMQ** [comment]
This was happening to me and I was able to fix by not running dataset.cache() before training

**Comment by sushreebarsa** [comment]
@Mageswaran1989 Could you try as per the above comment with latest stable version of TF 2.5 and let us know if the issue still pesists? Thanks!

**Comment by ttang235** [comment]
> same issue, TF 2.0
> I run Estimator with distributed mode on 6 machines ( PS x 1 / evaluator x 1 / worker x 4) ,
> my code is super simple：
> 
> ```
> model = tf.estimator.DNNLinearCombinedClassifier()
> train_spec = tf.estimator.TrainSpec(...) # input_fn: epoches = infinite
> eval_spec  = tf.estimator.EvalSpec(...) # input_fn: epoches = 1 
> tf.estimator.train_and_evaluate(model, train_spec, eval_spec)
> ```
> 
> As `save_checkpoints_steps = 500` , so it will generate a checkpoint every 500 steps, and the evaluator starts to evaluate when there is a new checkpoint. I found that each evaluation will result in an increase of about 500MB memory used on the evaluator node and other nodes (PS & Worker) did not change significantly.
> 
> PS: **Even if I use very small datasets when evaluating, memory leaks still exist. (~500MB↑/run)**

Have you figured out why this happened? We have the same issue. We also used train_and_evaluate with ParameterServerStrategy. In RunConfig we set eval_distribute=None (we used to set eval_distribute to ParameterServerStrategy in tf 1.15 and it worked well, but it won't work in tf 2.5.0), not sure whether it's related.

**Comment by arianmaghsoudnia** [comment]
I have the same problem in TF 2.8.0

**Comment by no1nemo** [comment]
I am having the same issue when running the trainer.py from the keypose repo from Google. [https://github.com/google-research/google-research/blob/master/keypose/trainer.py] I am running TF 2.8.0 and in conda

**Comment by Jiangyu1181** [comment]
> I experience the same issue on 2.4.0.

 I experience the same issue on 2.6.0.
